### PR TITLE
cmake: Re-organize syscall generation wrt. the build system

### PR DESCRIPTION
**Pull request by SebastianBoe**
This commit fixes
https://github.com/zephyrproject-rtos/zephyr/issues/5008.

It does so by splitting up gen_syscalls.py into two scripts with a
json metadata file to communicate syscall metadata between them. The
parsing script parses header files from include/ and writes syscall
metadata to a file if the metadata changed. The generation script
reads from the json file and generates syscall code.

The build system DAG now looks like this:

always_rebuild -> json -> syscalls -> offset.o

The script for generating json will do so only if the content changes,
this ensures that the entire DAG does not always do a full rebuild.

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

### CODE REVIEW

File: scripts/parse_syscalls.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    args = parser.parse_args()
+
+def main():
+    parse_args()
+
+    syscalls = analyze_headers(args.include)
+    
+    syscalls_in_json = json.dumps(syscalls, sort_keys=True)
+
+    # Check if the file already exists, and if there are no changes,
```
Comment (lpereira) [inline scripts/parse_syscalls.py]: Do we still need to do this check?

Comment (SebastianBoe) [reply scripts/parse_syscalls.py]: Yes. Without it there would be a timestamp update that propogates through the system.

**Review by tyrantlizard77 — COMMENTED** [review]
Instead of splitting the system call script in half, maybe we could do a simpler approach?

Right now, gen_syscalls.py unconditionally updates the file in the --syscall-dispatch argument, emits the syscall list header to stdout, and then conditionally updates the generated headers. It relied on the old build system's filechk logic to not touch the dispatch table or list header if it wasn't changed.

We could make this simpler:
1) Add an argument for the path to the system call list header, instead of emitting the data to stdout.
2) Add checks if the file has changed for the dispatch table and list header, just like how we do with the generated syscall include files
3) Run the script unconditionally on every build, but if nothing has changed, nothing should be touched by the script.

**Comment by SebastianBoe** [comment]
This works assuming that all of the information that we would want to be able to trigger a rebuild is
in the dispatch table and the list header.

If you can confirm that this is so then I'll rewrite this as proposed.

syscalls.json has entries that look like this:

```
    [
        "kernel.h",
        "_handler_k_object_access_grant",
        "K_SYSCALL_DECLARE2_VOID(K_SYSCALL_K_OBJECT_ACCESS_GRANT, k_object_access_grant, void *, object, struct k_thread *, thread);",
        "K_SYSCALL_K_OBJECT_ACCESS_GRANT",
        "[K_SYSCALL_K_OBJECT_ACCESS_GRANT] = _handler_k_object_access_grant"
    ],

```

But I see that the following information in syscalls.json but not in sycall_dispatch.c or syscall_list.h.

The name of the header file that the syscall comes from.
Type information, specifically, this entry:

K_SYSCALL_DECLARE2_VOID(K_SYSCALL_K_OBJECT_ACCESS_GRANT, k_object_access_grant, void *, object, struct k_thread *, thread);",

Stopping a regeneration when the type information changes would be a problem wouldn't it @andrewboie ?

I appreciate trying to find a simpler solution, but one could actually argue that splitting up the parsing
and generation of syscall metadata makes the system simpler as well ...

The syscall metadata is now located in an easy to access machine and human-readable format
so it is easy to have multiple consumers of syscall metadata.

**Comment by SebastianBoe** [comment]
@andrewboie : Your proposal is pretty straight forward, I could implement it and test what happens
when type information changes if this seems unclear.

**Comment by andrewboie** [comment]
gen_syscall_header.py has the same problem, I modified it and rebuilt, but the generated header was not updated

**Comment by SebastianBoe** [comment]
This PR has now bitrotted.

@andrewboie ; I believe your proposal for simplifying was flawed, please ACK or NACK this proposal and then I'll fix the bitrot.

**Comment by andrewboie** [comment]
ACK

**Comment by andrewboie** [comment]
Can you be sure to update the documentation if these changes contradict it?
http://docs.zephyrproject.org/kernel/usermode/syscalls.html

**Comment by SebastianBoe** [comment]
> Can you be sure to update the documentation if these changes contradict it?

Done.

**Comment by SebastianBoe** [comment]
This is ready to be approved and merged.

**Review by dbkinder — APPROVED** [review]
+1 for docs

**Comment by galak** [comment]
@andrewboie can you please review

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**